api/mesh/v1alpha1/gateway.proto

```diff
@@ -0,0 +1,146 @@
+syntax = "proto3";
+
+package kuma.mesh.v1alpha1;
+
+option go_package = "github.com/kumahq/kuma/api/mesh/v1alpha1";
+
+import "mesh/options.proto";
+import "mesh/v1alpha1/selector.proto";
+import "validate/validate.proto";
+
+// Gateway is a virtual proxy.
+//
+// Each Gateway is bound to a set of builtin gateway dataplanes.
+// Each builtin dataplane instance can host exactly one Gateway
+// proxy configuration.
+//
+// Gateway aligns with the Kubernetes Gateway API v1alpha2. See that
+// spec for detailed documentation.
+message Gateway {
+
+  option (kuma.mesh.resource).name = "GatewayResource";
+  option (kuma.mesh.resource).type = "Gateway";
+  option (kuma.mesh.resource).package = "mesh";
+
+  option (kuma.mesh.resource).skip_validation = true;
+  option (kuma.mesh.resource).skip_registration = true;
+
+  enum Protocol {
+    PROTO_NONE = 0;
+    PROTO_TLS = 1;
+    PROTO_HTTP = 2;
+    PROTO_HTTPS = 3;
+  }
+
+  // Aligns with RouteBindingSelector.
+  message RouteSelector {
+    // Type of resources to select.
+    string type = 1;
+
+    // Resource tags to match. This selects all the resources of the
+    // given type that have tags that match the selector.
+    Selector tags = 2;
+
+    // XXX(jpeach) For Kubernetes Gateway, we need to make this a oneof
+    // union, where one branch of the union has group, kind, namespace
+    // and selector.
+  }
+
+  // Aligns with LocalObjectReference.
+  message ObjectReference {
+    // Type of type resource that is being referenced.
+    string type = 1;
+    // Name of the resource that is being referenced.
+    string name = 2;
+
+    // XXX(jpeach) For Kubernetes Gateway, we need to make this a oneof
+    // union, where one branch of the union has group, kind, and name.
+  }
+
+  // TLSConfig describes a TLS configuration.
+  message TLS {
+    enum Mode {
+      TLS_NONE = 0;
+      // The TLS session between the downstream client and the Gateway
+      // is terminated at the Gateway. This mode requires certificateRef
+      // to be set.
+      TLS_TERMINATE = 1;
+      // The TLS session is NOT terminated by the Gateway. This implies
+      // that the Gateway can't decipher the TLS stream except for the
+      // ClientHello message of the TLS protocol. The certificateRef field
+      // is ignored in this mode.
+      TLS_PASSTHRU = 2;
+    }
+
+    // Aligns with GatewayTLSConfig.
+    message Conf {
+      // Mode defines the TLS behavior for the TLS session initiated by the
+      // client.
+      Mode mode = 1;
+
+      // CertificateRef is a reference to a object that contains a TLS
+      // certificate and private key. This certificate is used to establish
+      // a TLS session for requests that match the hostname of the associated
+      // listener.
+      ObjectReference certificateRef = 2;
+
+      // Options are a list of key/value pairs to give extended options
+      // to the provider.
+      map<string, string> options = 4;
```

Comment:
Per Jakub we could use a typed options message (empty for now) and convert the Gateway  API map into that.